Metrics fo realz

jaredhirsch opened the pull request:
**Update: no longer work in progress, this bad boy is ready to merge**

---

WIP commit, still needs a little cleanup. @chuckharmston mind taking a look?

Opening as WIP to make sure we're gathering enough data for the first pass. Comments welcome.

jaredhirsch committed:
Implement minimal version of metrics

* Update metrics documentation to reflect reality.

* Assign the RecommendationRow to the Recommendation, so the metrics
  ping can minimize hitting the DOM to get information.

* Fire an event when the popup is about to open, so the metrics code can
  start listening for events.

* Fire an event when the popup detects a click, so the metrics code can
  get an accurate selectedIndex value.

Fixes #18.

jaredhirsch commented:
One outstanding TODO: this needs to be rebased against master, and we need to include the recommendation type (wikipedia vs tld) in the packet that's sent down.

jaredhirsch commented on lib/metrics.js:
```diff
@@ -0,0 +1,129 @@
+
+      // What was the index of the selected item?
+      // If the recommendation was selected, this will be -1.
+      selectedIndex: null
```
@nchapman Does this look like it answers the basic product questions? This packet is constructed each time the popup opens, and is fired each time the popup closes, regardless of whether the user navigated or interacted with the popup while it was open. One missing bit of data that I'll add is the wikipedia vs TLD type.

This data should be enough for us to calculate the click-through rate of the popup with and without the recommendation visible, and drill down into click-through rates of the different types of recommendations.

We should also be able to figure out how often users interact with the popup, since it always pops open when you enter stuff in the urlbar.

nchapman replied:
The two [KPIs identified](https://github.com/mozilla/universal-search/blob/master/docs/metrics.md#key-performance-indicators) are:
- the click through rate of the recommendations
- retention (how do people use recommendations over time)

and I think this let's us answer both those questions, so lgtm!

jaredhirsch commented:
@lmorchard @clouserw This metrics patch is coming together, comments welcome if you have spare cycles. The way I'm doing it is fast and simple, but will require writing some code to calculate KPI percentages on the server. Since we'll have some flexibility in data visualization, it seems like this should work fine.

jaredhirsch commented on lib/metrics.js:
```diff
@@ -0,0 +1,129 @@
+function Metrics(opts) {
+  /*
+  The metrics class is used to determine whether the recommendation is used
+  when the user navigates with the popup open. Data is sent to the Test Pilot
```
This is not quite accurate. We fire a signal every time the popup closes, whether or not the user navigated or hit `Escape` to dismiss the popup.

jaredhirsch commented on lib/metrics.js:
```diff
@@ -0,0 +1,129 @@
+
+Metrics.prototype = {
+  init: function() {
+    this.win.console.log('metrics.init');
```
All of the console.log() calls will be removed when this graduates from WIP to real thing

chuckharmston replied:
We should probably consider a unified logging approach, though, to help facilitate error reporting. something like:

```
Logger.prototype = {
  _prefix: function() {
    return `[${this.classPrefix}.${this.instancePrefix}]`;
  },

  log: function() {
    [].splice.apply(arguments, [0, 0, this._prefix()]);
    this.win.console.log.apply(this.win.console, arguments);
  },

  warn: function() {
    [].splice.apply(arguments, [0, 0, this._prefix()]);
    this.win.console.warn.apply(this.win.console, arguments);
  },

  error: function() {
    [].splice.apply(arguments, [0, 0, this._prefix()]);
    this.win.console.error.apply(this.win.console, arguments);
  }
};


// Then in the Metrics class...
this.logger = new this.Logger('metrics', this.win);

// Then in Metrics.init, to log the Metrics instance:
this.logger.log('init', this);

// Console sez:
'[universal-search.metrics]' 'init' Object { ... }
```

jaredhirsch replied:
Sending logger output to Telemetry sounds like a good idea, could help us figure out how frequently users hit errors. Maybe we can file a followup bug?

chuckharmston replied:
https://github.com/mozilla/universal-search/issues/184 :)

I'm working on this now so I can add some useful logging for @clouserw for #181.

6a68 commented on lib/metrics.js:
```diff
@@ -0,0 +1,129 @@
+    this.checked = false;
+    this.interaction = {
+
+      // Did the user navigate?
```
Better: "Did the popup close because the user navigated to a URL?"

jaredhirsch commented on lib/metrics.js:
```diff
@@ -0,0 +1,129 @@
+      recommendationType: this.recommendation.resultType,
+      // If no selectedIndex was passed in, we get the value from the popup.
+      // Note that `isNaN(undefined) => true`.
+      selectedIndex: isNaN(selectedIndex) ? this.popup.el.selectedIndex : selectedIndex
```
Hm. It's possible that the user might hit a letter and the enter key faster than the popup can open for the first time. In that case, `this.popup.el` might not exist yet, so that this would throw.

jaredhirsch commented on lib/metrics.js:
```diff
@@ -0,0 +1,129 @@
+      const data = {
+        id: 'universal-search@mozilla.com',
+        data: JSON.stringify(this.interaction)
+      };
```
oops, this needs to be { subject, data } https://github.com/mozilla/testpilot/blob/master/addon/lib/metrics.js#L111

jaredhirsch commented on lib/ui/recommendation.js:
```diff
@@ -167,6 +167,12 @@ Recommendation.prototype = {
+      click: !!evt.button,
+      type: this.resultType
+    };
+
```
The changes in this file are stale (from an earlier attempt), ignore. I'll remove.

chuckharmston commented:
Metrics spec, for reference: https://github.com/mozilla/universal-search/blob/master/docs/metrics.md

chuckharmston commented on lib/metrics.js:
```diff
@@ -0,0 +1,129 @@
+}
+
+Metrics.prototype = {
+  init: function() {
```
This ended up being much cleaner and less intrusive than I thought it would thanks to how you've architected this. Really nicely done.

jaredhirsch replied:
:beers: Yeah, unobtrusive measurement is one of the nicest things about extremely late binding via pubsub (unit testing is another). If we had modules calling each other directly, or used promise-based APIs, it'd be a lot messier.

chuckharmston commented on lib/metrics.js:
```diff
@@ -0,0 +1,129 @@
+      recommendationShown: false,
+
+      // What type of recommendation was it?
+      recommendationType: null,
```
Notably, we also want a list of all the recommendation's enhancements `Object.keys(recommendation.enhancements)` if a recommendation is selected.

jaredhirsch replied:
How about we file a followup bug for adding enhancements after public launch?

Based on the [examples](https://gist.github.com/chuckharmston/861a1442e7dde98aefe333a2347e0361) we discussed the other day, it looks like every TLD entry will have wikipedia, tld, favicon enhancements, while wikipedia entries will have wikipedia, favicon enhancements. This variation is captured by the recommendationType.

chuckharmston replied:
That's not new scope, it's covered here: https://github.com/mozilla/universal-search/blob/master/docs/metrics.md#on-the-client

> List of enhancers applied (e.g. `Object.keys(recommendation.enhancements)`.

Notably, not every TLD entry will have a wikipedia enhancement—only if that search term has a Wikipedia page in the first set of results.

chuckharmston replied:
`Notably` has apparently become a crutch word of mine.

chuckharmston commented:
Do we have the capacity to determine the type of non-recommendation selections? e.g. bookmark vs history vs search vs suggested search term? That was a pretty [key part](https://github.com/mozilla/universal-search/blob/master/docs/metrics.md#result-types) of the initial idea, helping us understand how users interact with the awesome bar, in addition to our recommendation.

chuckharmston commented:
We also want to collect information about [which data is shown](https://github.com/mozilla/universal-search/blob/master/docs/metrics.md#rich-data). With the recent change to distinct result types, we may want to tweak what exactly this reports.

chuckharmston commented:
This is brilliant. Thank you!

jaredhirsch commented:
Thanks for the feedback, @chuckharmston! There's definitely room to improve these metrics quite a bit, and I think we should file bugs for the first sprint post-public launch. This is intended to be a (very minimal) minimal viable metrics patch.

The data gathered by this patch lets us answer the fundamental questions product wants to answer:

**What is the click-through rate (CTR) of the popup overall?**
Given by: (events with `didNavigate = true`) / (total event count)

**What is the CTR when recommendations are shown?**
(events with `didNavigate = true` and `recommendationShown = true`) / (total events with `recommendationShown = true`)

**CTR when recommendations are not shown?**
Same as previous, with `recommendationShown = false`

**CTR when wikipedia recommendations are shown?**
(`didNavigate = true` and `recommendationType = wikipedia`) / (total events with `recommendationType = wikipedia`)

**CTR when TLD recommendations are shown?**
Same as previous, just swap out the `recommendationType`

**When users navigate via the popup, how often do they use the mouse?**
(`didNavigate = true` and `interactionType = click`) / (total events with `didNavigate = true`)

**When users navigate via the popup, how often do they use the keyboard?**
Same as previous, just swap out the `interactionType`.

**CTR for the 3rd result down the list?**
(`didNavigate = true` and `selectedIndex == 2`) / (total events with `didNavigate = true`)

I definitely want to instrument this more thoroughly, but I think this should be good enough for a first pass. cc @nchapman

_Edited to add CTR without recommendations shown and measure the selected index_

nchapman commented:
@6a68 this looks good! I commented, inline above, that as long as we can measure our 2 KPIs (CTR and retention) then we're good to go, and that very much appears to be true. We can add the secondary metrics in another pass.

jaredhirsch commented:
Great, I'll get this PR cleaned up and remove the "WIP" bit from the title, and we can land it today.

jaredhirsch committed:
Guard against recommendation.el being undefined

I'm not sure under what circumstances this happens, but I've seen the
error go by in the browser console.

jaredhirsch commented:
@chuckharmston I guess this is not quite ready to land. I'm trying to figure out how to verify if this is working or not, because it's not clear exactly what syntax will work with the `notifyObservers` call. Going to try adding some logging to the test pilot add-on to verify.

chuckharmston commented on docs/metrics.md:
```diff
@@ -16,36 +17,73 @@ Metrics gathered from the Universal Search add-on will be reported to the Test P
 
+- The popup usually opens when the user types in the urlbar. It also opens in some less common circumstances, like clicking the history drop marker (the small down caret to the right of the urlbar).
+- The popup closes when the user navigates or dismisses the popup without navigating.
+  - The popup is dismissed by hitting the 'Escape', left arrow, or right arrow keys.
```
~~Also `enter`~~

Maybe `can be dismissed without navigating by`?

jaredhirsch replied:
:+1:

chuckharmston commented on docs/metrics.md:
```diff
@@ -16,36 +17,73 @@ Metrics gathered from the Universal Search add-on will be reported to the Test P
+	"test": "universal-search@mozilla.com",  // The em:id field from the add-on
+	"agent": "User Agent String",
+	"payload": {
+		“didNavigate”: true,
```
Curly quotes on the key here.

jaredhirsch replied:
I have no idea how I did that. Trying again.

chuckharmston commented on docs/metrics.md:
```diff
@@ -16,36 +17,73 @@ Metrics gathered from the Universal Search add-on will be reported to the Test P
+
+#### Data Analysis
+
+We will use this data to calculate click-through rates for the popup as shown below.
```
Nit: add `(CTR)` after `click-through rate`

jaredhirsch replied:
:+1:

chuckharmston commented:
Some documentation nits, but this looks great. 

> Going to try adding some logging to the test pilot add-on to verify.

If you do that and it seems to work, :shipit:!

jaredhirsch commented:
![](https://media.giphy.com/media/vv41HlvfogHAY/giphy.gif)

chuckharmston commented:
> I'm not sure under what circumstances this happens, but I've seen the error go by in the browser console.

I think this is just a timing error; because of the nature of XBL, it doesn't get inserted until the popup until the popup is instantiated for the first time, so any event-driven code that relies on it has to account for the fact that it may not quite be there. It's pretty easy to workaround, since it's an acceptable experience to just not show a recommendation.
